chain/network/src/concurrency/scope/mod.rs

```diff
@@ -0,0 +1,459 @@
+    }
+}
+
+/// Wrapper of a scope reference which cancels the scope when dropped.
+///
+/// Used by Scope to cancel the scope as soon as all tasks spawned via
+/// `Scope::spawn` complete.
+struct StrongService<E: 'static>(Arc<Inner<E>>);
+
+impl<E: 'static> Borrow<Inner<E>> for StrongService<E> {
+    fn borrow(&self) -> &Inner<E> {
+        &*self.0
+    }
+}
+
+impl<E: 'static> Drop for StrongService<E> {
+    fn drop(&mut self) {
+        self.0.ctx.cancel()
+    }
+}
+
+/// Scope represents a concurrent computation bounded by lifetime 'env.
+///
+/// It should be created only via `run!` macro.
+/// Scope is cancelled when the provided context is cancelled.
+/// Scope is cancelled when any of the tasks in the scope returns an error.
+/// Scope is cancelled when all the tasks in the scope complete.
+/// Scope is terminated when it is cancelled AND all tasks in the scope complete.
+pub struct Scope<'env, E: 'static>(
+    /// Scope is equivalent to a strong service, but bounds
+    Weak<StrongService<E>>,
+    Weak<Inner<E>>,
+    /// Makes Scope<'env,E> invariant in 'env.
+    std::marker::PhantomData<fn(&'env ()) -> &'env ()>,
+);
+
+unsafe fn to_static<'env, T>(f: BoxFuture<'env, T>) -> BoxFuture<'static, T> {
+    std::mem::transmute::<BoxFuture<'env, _>, BoxFuture<'static, _>>(f)
+}
+
+impl<'env, E: 'static + Send> Scope<'env, E> {
+    /// Spawns a "main" task in the scope.
+    /// Scope gets canceled as soon as all the "main" tasks complete.
+    pub fn spawn<T: 'static + Send>(
+        &self,
+        f: impl 'env + Send + Future<Output = Result<T, E>>,
+    ) -> JoinHandle<'env, T> {
+        match self.0.upgrade() {
+            Some(strong) => JoinHandle(
+                Inner::spawn(strong, unsafe { to_static(f.boxed()) }),
+                std::marker::PhantomData,
+            ),
+            // Upgrade may fail only if all the "main" tasks have already completed
+            // so the caller is a "background" task. In that case we fall back
+            // to spawning a "background" task instead. It is ok, since the distinction
+            // between main task and background task disappears, once the scope is canceled.
+            None => self.spawn_bg(f),
+        }
+    }
+
+    /// Spawns a "background" task in the scope.
+    /// It behaves just like a single-task Service, but
+    /// has the same lifetime as the Scope, so it can spawn
+    /// more tasks in the scope. It is not a "main" task, so
+    /// it doesn't prevent scope cancelation.
+    pub fn spawn_bg<T: 'static + Send>(
+        &self,
+        f: impl 'env + Send + Future<Output = Result<T, E>>,
+    ) -> JoinHandle<'env, T> {
+        JoinHandle(
+            Inner::spawn(self.1.upgrade().unwrap(), unsafe { to_static(f.boxed()) }),
+            std::marker::PhantomData,
+        )
+    }
+
+    /// Spawns a service.
+    ///
+    /// Returns a handle to the service, which allows spawning new tasks within the service.
```

Comment:
The docs should mention that this will panic if called outside a task spawned by this `Scope`, if I understand your answer to @nagisa ’s comment correctly.

Reply:
The point is that as long as the only way of constructing the Scope is via run! macro, the weak pointers are actually non-null throughout the whole run() call (an Arc which guarantees that is kept alive in every task of the scope separately). See https://github.com/near/nearcore/pull/8647#discussion_r1124579433 for why it cannot be an Arc.